File: NBE_5_7_2_02TEAM/src/main/java/io/twogether/nbe_5_7_2_02team/post/domain/RecruitmentField.java
package io.twogether.nbe_5_7_2_02team.post.domain;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;

import lombok.AccessLevel;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;

@Entity
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class RecruitmentField {

    public String getFieldName() {
        return fieldName;
    }

    public boolean isClosed() {
        return closed;
    }

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "post_id")
    private Post post;

    @Column(nullable = false)
    private String fieldName;

    @Column(nullable = false)
    private int totalCount;

    @Column(nullable = false)
    private int currentCount = 0;

    @Column(nullable = false)
    private boolean closed = false;

    @Builder
    public RecruitmentField(
            Post post, String fieldName, int totalCount, int currentCount, boolean closed) {
        this.post = post;
        this.fieldName = fieldName;
        this.totalCount = totalCount;
        this.currentCount = currentCount;
        this.closed = closed;
    }

    public void close() {
        this.closed = true;
    }

    public void increaseCount() {
        this.currentCount++;
        if (this.currentCount >= this.totalCount) {
            this.closed = true;
        }
    }
}

File: NBE_5_7_2_02TEAM/src/main/java/io/twogether/nbe_5_7_2_02team/post/domain/RecruitmentField.kt
package io.twogether.nbe_5_7_2_02team.post.domain

import jakarta.persistence.*
import jakarta.persistence.Column
import lombok.AccessLevel
import lombok.Builder
import lombok.NoArgsConstructor

@Entity
class RecruitmentField (
    @field:JoinColumn(name = "post_id")
    @field:ManyToOne(
        fetch = FetchType.LAZY
    ) var post: Post?,

    @field:Column(nullable = false)
    var fieldName: String,

    @field:Column(nullable = false)
    var totalCount: Int,

    @field:Column(nullable = false)
    var currentCount: Int,

    @field:Column(nullable = false)
    var closed: Boolean
) {
    @field:Id
    @field:GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: Long? = null

    fun increaseCount() {
        currentCount++
        if (currentCount >= totalCount) {
            closed = true
        }
    }
}
